cpp/src/KIM_FilesystemPath.hpp

```diff
@@ -0,0 +1,203 @@
+{
+ public:
+  Path() {}
+  Path(const char * str) : path_(str) {}
+  Path(const std::string & str) : path_(str) {}
+#ifdef KIM_API_USE_FILESYSTEM_LIBRARY
+  Path(const std::filesystem::path & p) : path_(p) {}
+  Path(std::filesystem::path && p) : path_(std::move(p)) {}
+#endif
+
+  Path & operator=(const std::string & other)
+  {
+    path_ = other;
+    return *this;
+  }
+
+  Path & operator=(const char * other)
+  {
+    path_ = other;
+    return *this;
+  }
+
+  Path & operator+=(char const * const s);
+  Path & operator/=(const Path & p);
+  Path operator/(const Path & p) const;
+  friend bool operator<(const Path & lhs, const Path & rhs)
+  {
+    return lhs.path_ < rhs.path_;
+  }
+  friend bool operator==(const Path & lhs, const Path & rhs)
+  {
+    return lhs.path_ == rhs.path_;
+  }
+
+  // Concatenates the current path and the argument.
+  Path & concat(const std::string & p);
+
+  // Turns the path object into a conventional string, which can be passed to
+  // I/O functions.
+#ifdef KIM_API_USE_FILESYSTEM_LIBRARY
+  std::string string() const { return path_.string(); }
+#else
+  const std::string & string() const { return path_; }
+#endif
+
+  // Resets the path to an empty string.
+  void clear() { path_.clear(); }
+
+  // Returns whether this path is the empty string.
+  bool empty() const { return path_.empty(); }
+
+  // Returns the last component of the path.
+  Path filename() const;
+
+  // Removes a single filename component from the path.
+  Path & remove_filename();
+
+  // Converts all directory separators to the preferred directory separator of
+  // the current platform.
+  Path & make_preferred();
+
+  // Creates a new directory, including parent directories if necessary.
+  // It's not an error if the directory to be created already exists.
+  // Returns true on error.
+  bool make_directory() const;
+
+  // Deletes the contents of this path (if it is a directory) and the contents
+  // of all its subdirectories, recursively, then deletes the file path itself.
+  // Returns true on error.
+  bool remove_directory_recursive() const;
+
+  // Returns the list of subdirectories of this directory.
+  std::vector<Path> subdirectories() const;
+
+  // Checks whether the file or directory exists.
+  bool exists() const;
+
+  // Checks whether the path is relative.
+  bool is_relative() const;
+
+  // Returns the current working directory.
+  static Path current_path();
+
+  // Returns the user's home directory.
+  static Path home_path();
+
+  // Creates a new empty directory that can be used to write temporary files
+  // into. Returns an empty path on failure.
+  static Path create_temporary_directory(char const * const namePrefix);
+
+  // Performs stream output on the path (operator <<).
+  template<class CharT, class Traits>
+  friend std::basic_ostream<CharT, Traits> &
+  operator<<(std::basic_ostream<CharT, Traits> & os, const Path & p)
+  {
+    return os << p.path_;
+  }
+
+#ifndef KIM_API_USE_FILESYSTEM_LIBRARY
```

Comment:
This can be defined in the `.cpp` and the `#ifdef` avoided.

Reply:
Please remove the surrounding `#ifdef` here. `preferred_separator` should always be defined. This wasn't the case before when `Path` inherited this field from the `std::filesystem::path` base class in C++17 mode.